### DISCUSSION TITLE
macOS gamepad support

### DISCUSSION THREAD

**Discussion by kotleni** [LizardByte/community#130] [Sunshine] [answered]
https://github.com/LizardByte/Sunshine/blob/08d623ce44f77061a26de9dc50a2aef60f1b6158/src/platform/macos/input.cpp#L290
_What about gamepad support on macOS? Is this worth the wait?_

**Comment by kotleni** [comment]
I started developing gamepad support on macOS. This is already working great.
After the refactoring, I plan to make a Pull Request.

<img width="1013" alt="Screenshot 2023-01-11 at 19 09 42" src="https://user-images.githubusercontent.com/38311102/211872734-407b4c49-9059-4d95-bb49-b52229eda738.png">

**Reply by ReenigneArcher** [reply]
Thank you, this will be a great addition to Sunshine! Please make your PR against our `nightly` branch.

**Reply by kotleni** [reply]
> Thank you, this will be a great addition to Sunshine! Please make your PR against our `nightly` branch.

Okay. But first I need fix some bugs. (buttons have wrong mapping)

**Reply by kotleni** [reply]
Done

**Reply by deanezra** [reply]
@kotleni did you put up a pull request in the end? I would love to build and try this out on my mac m2

**Reply by ReenigneArcher** [reply]
@deanezra It's not really useable, but there is the following:

- https://github.com/LizardByte-research/VirtualHID-macOS
- https://github.com/LizardByte/Sunshine/pull/756

**Comment by Githsams** [comment]
Hello, is gamepad working on sunshine for mac now ?

**Reply by ReenigneArcher** [reply]
No

**Reply by kotleni** [reply]
Not yet, not yet... 😞

**Comment by actstorms** [comment]
I hope this will be integrated.

**Comment by Rickyr921** [comment]
Any luck on this? Been checking this thread regularly in hopes that I won’t need to buy a secondary machine for only streaming games when my Mac already runs them well. 🙏🙏🙏

**Reply by ReenigneArcher** [reply]
No one is working on this.

**Comment by ramirocantu** [comment]
I'd like to try working on this. any pointers would be helpful.

I found the game controller documentation on [Apple's website](https://developer.apple.com/documentation/gamecontroller). I can try my hand at this this week.

I'm a highschool teacher and usually pretty busy and I dont have much experience with mac development, but I hope that this can be pretty doable and a good way to get my feet wet.

**Reply by Hazer** [reply]
@ramirocantu I don't think game controller docs are what you're looking for. The Game Controller API is for an app to receive and use the inputs, but sunshine receives them from the network. Sunshine needs to actually execute the inputs to the OS host, so games and apps can receive, and the GameController API can't execute to OS, just the App that it is running will receive events.

I believe what you're looking for is actually [DriverKit](https://developer.apple.com/documentation/driverkit), and doing something similar to [noah-nuebling/mac-gamepad-driver](https://github.com/noah-nuebling/mac-gamepad-driver), but instead of supporting physical xbox360 controllers, supporting a virtual controller that Sunshine implements.

**Reply by riiswa** [reply]
@ramirocantu did you start working on it? I can also help if you need :)

**Reply by kotleni** [reply]
@ramirocantu @riiswa https://github.com/LizardByte/Sunshine/pull/756

**Reply by riiswa** [reply]
@kotleni what was your VirtualHID ?

**Reply by ReenigneArcher** [reply]
I forked it before it was deleted: https://github.com/LizardByte-research/VirtualHID-macOS

**Comment by J7mbo** [comment]
I hope this will be integrated either this or next year. Following and I'll come back in two years.

**Reply by ReenigneArcher** [reply, marked as answer]
See https://github.com/LizardByte/roadmap/issues/21

Note: I am marking this as the answer to this discussion, though it does not mean the feature is integrated yet.
